### PR TITLE
Db api

### PR DESCRIPTION
**Pull request by dpsutton**
Let's get an api namespace for `metabase.db` and company.

Namespaces in the `src/metabase/db` folder:

```
❯ tree src/metabase/db
src/metabase/db
├── connection.clj
├── connection_pool_setup.clj
├── custom_migrations.clj
├── data_source.clj
├── env.clj
├── jdbc_protocols.clj
├── liquibase
│   ├── h2.clj
│   └── mysql.clj
├── liquibase.clj
├── metadata_queries.clj
├── query.clj
├── setup.clj
├── spec.clj
└── update_h2.clj

2 directories, 14 files
```

This ends up with two namespaces:
- `metabase.db`
- `metabase.db.query`: this seems to be pretty extensively used and is a quasi api namespace in its own right. I don't have a better name for it at the moment

##### metabase.db
```clojure
db=> (dir metabase.db)
app-db
data-source
db-file
db-is-set-up?
db-type
increment-app-db-unique-indentifier!
make-subname
memoize-for-application-db
migrate!
quoting-style
recent-activity?
setup-db!
spec
unique-identifier
```

##### metabase.db.query
```clojure
db=> (dir metabase.db.query)
compile
format-sql
isa
join
qualify
query
reducible-query
select-or-insert!
update-or-insert!
with-conflict-retry
```

#### src/metabase/cmd

This does some special one-offs with app-db connections for commands and establishing a app-db connection that felt a bit more specialized than we want the api to be.

#### src/metabase/driver/h2.clj

Does some special registering for clob reading. Could probably be tidied up but would prefer someone from the drivers team to do this.

#### metabase.db.metadata-queries

Used in some sync related stuff. This could be moved to a sync namespace in the future. This was getting a bit larger so I'll delay that.

### Finding usages:

Some helpful utilities i used:

Finding all `[metabase.db.<something> :as <something>]` references:
```
ag --ignore src/metabase/db --ignore src/metabase/db.clj "\[metabase\.db\..*\ :as .*]" src enterprise > usages

cat usages
src/metabase/cmd/dump_to_h2.clj:20:   [metabase.db.connection :as mdb.connection]
src/metabase/cmd/copy.clj:12:   [metabase.db.setup :as mdb.setup]
src/metabase/cmd/copy/h2.clj:6:   [metabase.db.data-source :as mdb.data-source]
src/metabase/metabot/util.clj:9:   [metabase.db.query :as mdb.query]
src/metabase/driver/h2.clj:8:   [metabase.db.jdbc-protocols :as mdb.jdbc-protocols]
src/metabase/driver/sql_jdbc/sync/describe_table.clj:9:   [metabase.db.metadata-queries :as metadata-queries]
src/metabase/models/bookmark.clj:5:   [metabase.db.query :as mdb.query]
src/metabase/models/collection/graph.clj:6:   [metabase.db.query :as mdb.query]
src/metabase/models/table.clj:4:   [metabase.db.query :as mdb.query]
src/metabase/models/pulse_channel.clj:6:   [metabase.db.query :as mdb.query]
src/metabase/models/permissions_group_membership.clj:3:   [metabase.db.query :as mdb.query]
src/metabase/models/params.clj:16:   [metabase.db.query :as mdb.query]
src/metabase/models/card.clj:14:   [metabase.db.query :as mdb.query]
src/metabase/models/params/field_values.clj:6:   [metabase.db.query :as mdb.query]
src/metabase/models/params/chain_filter.clj:70:   [metabase.db.query :as mdb.query]
src/metabase/models/field_values.clj:30:   [metabase.db.query :as mdb.query]
src/metabase/models/permissions_group.clj:13:   [metabase.db.query :as mdb.query]
src/metabase/models/dashboard.clj:9:   [metabase.db.query :as mdb.query]
src/metabase/models/database.clj:4:   [metabase.db.query :as mdb.query]
src/metabase/models/moderation_review.clj:5:   [metabase.db.query :as mdb.query]
src/metabase/models/user.clj:7:   [metabase.db.query :as mdb.query]
src/metabase/models/dashboard_card.clj:6:   [metabase.db.query :as mdb.query]
src/metabase/automagic_dashboards/core.clj:159:   [metabase.db.query :as mdb.query]
src/metabase/task/sync_databases.clj:14:   [metabase.db.query :as mdb.query]
src/metabase/api/activity.clj:7:   [metabase.db.query :as mdb.query]
src/metabase/api/collection.clj:16:   [metabase.db.query :as mdb.query]
src/metabase/api/table.clj:8:   [metabase.db.query :as mdb.query]
src/metabase/api/database.clj:12:   [metabase.db.query :as mdb.query]
src/metabase/api/permissions.clj:12:   [metabase.db.query :as mdb.query]
src/metabase/api/field.clj:6:   [metabase.db.metadata-queries :as metadata-queries]
src/metabase/api/field.clj:7:   [metabase.db.query :as mdb.query]
src/metabase/api/public.clj:16:   [metabase.db.query :as mdb.query]
src/metabase/api/search.clj:10:   [metabase.db.query :as mdb.query]
src/metabase/sync/analyze/fingerprint.clj:7:   [metabase.db.metadata-queries :as metadata-queries]
src/metabase/sync/analyze/fingerprint.clj:8:   [metabase.db.query :as mdb.query]
src/metabase/query_processor/middleware/cache_backend/db.clj:5:   [metabase.db.query :as mdb.query]
src/metabase/email/messages.clj:13:   [metabase.db.query :as mdb.query]
src/metabase/analytics/stats.clj:11:   [metabase.db.query :as mdb.query]
enterprise/backend/test/metabase_enterprise/serialization/test_util.clj:7:   [metabase.db.connection :as mdb.connection]
enterprise/backend/test/metabase_enterprise/serialization/test_util.clj:8:   [metabase.db.data-source :as mdb.data-source]
enterprise/backend/src/metabase_enterprise/sandbox/models/permissions/delete_sandboxes.clj:5:   [metabase.db.query :as mdb.query]
enterprise/backend/src/metabase_enterprise/audit_app/pages/common.clj:13:   [metabase.db.query :as mdb.query]
```

### CODE REVIEW

File: dev/src/dev.clj [hunk 1]
```diff
@@ -249,8 +248,8 @@
   ([]
    (migrate! :up))
   ([direction & [version]]
-   (mdb.setup/migrate! (mdb.connection/db-type) (mdb.connection/data-source)
-                       direction version)))
+   (mdb/migrate! (mdb/db-type) (mdb/data-source)
+                 direction version)))
```
Comment (tsmacdonald) [inline dev/src/dev.clj:252]: first file changed and it's already nicer reading to have one ns instead of two

File: src/metabase/db.clj [hunk 2]
```diff
@@ -1,49 +1,47 @@
...
-  * [[metabase.db.liquibase]] - high-level Clojure wrapper around relevant parts of the Liquibase API
-
-  * [[metabase.db.setup]] - code related to setting up the application DB -- verifying the connection and running
-    migrations -- and for setting it up as the default Toucan connection
-
-  * [[metabase.db.spec]] - util functions for creating JDBC specs for supported application DB types from connection
-    details maps
-
-  * [[metabase.db.util]] - general util functions for Toucan/HoneySQL queries against the application DB"
+  Other namespaces should only depend on this namespace and a sensible API for the app-db maintained."
```
Comment (tsmacdonald) [inline src/metabase/db.clj:9]: I'm having trouble parsing

> and a sensible API for the app-db maintained

File: src/metabase/db.clj [hunk 3]
```diff
@@ -1,49 +1,47 @@
...
-   [toucan2.pipeline :as t2.pipeline]))
+   [metabase.db.spec :as mdb.spec]
+   [potemkin :as p]))
+
+(set! *warn-on-reflection* true)
 
 ;; TODO - determine if we *actually* need to import any of these
 ;;
 ;; These are mostly here as a convenience to avoid having to rework a bunch of existing code. It's better to use these
 ;; functions directly where applicable.
```
Comment (tsmacdonald) [inline src/metabase/db.clj:24]:
```suggestion

```

I'd say you took care of that TODO!

File: src/metabase/db.clj [hunk 4]
```diff
@@ -1,49 +1,47 @@
...
+ [mdb.env
+  db-file]
+
+ [mdb.setup
+  migrate!]
+
+ [mdb.spec
+  make-subname
+  spec]
+ )
```
Comment (tsmacdonald) [inline src/metabase/db.clj:44]: do we really have dangling parens for potemkin imports? Would've expected this to be on the previous line

File: src/metabase/db.clj [hunk 5]
```diff
@@ -52,6 +50,12 @@
   []
   (= @(:status mdb.connection/*application-db*) ::setup-finished))
 
+(defn app-db
+  "The Application database. A record, but use accessors [[db-type]], [[data-source]], etc to acess. Also
```
Comment (tsmacdonald) [inline src/metabase/db.clj:54]:
```suggestion
  "The Application database. A record, but use accessors [[db-type]], [[data-source]], etc to access. Also
```

File: src/metabase/db/query.clj [hunk 6]
```diff
@@ -114,3 +175,87 @@
...
+                       ;; For convenience, we allow the update-fn to omit fields in the search-map
+                       (merge updated select-map)))]
+    (with-conflict-retry
+      (if-let [existing (apply t2/select-one model select-kvs)]
+        (let [pk      (pk-key existing)
+              updated (update-fn existing)]
+          (t2/update! model pk updated)
+          ;; the private key may have been changed by the update, and this is OK.
+          (pk-key updated pk))
+        (t2/insert-returning-pk! model (update-fn nil))))))
```
Comment (tsmacdonald) [inline src/metabase/db/query.clj:261]: surprised both of these fns aren't in toucan2 isntead of here, but that conversation is orthogonal to this PR

**Review by tsmacdonald — APPROVED** [review]
:boat:

File: src/metabase/db/query.clj (hunk 6 above)
Comment (crisptrutski) [reply src/metabase/db/query.clj:261]: [One day...](https://github.com/metabase/toucan/issues/17)

It's unclear that these are the semantics we'd want in such an agnostic library though. I think most people would expect to have a method that uses the atomic statement approach, which varies by database type. Or maybe it could provide both. 🤷

Comment (dpsutton) [reply src/metabase/db/query.clj:261]: 100% agree. Didn't know these existed until making an api and thinking "this doesn't belong in the api". But want to do this in stages. We can clean up the api in a follow up.

File: src/metabase/db/query.clj [hunk 7]
```diff
@@ -22,19 +22,80 @@
...
 (set! *warn-on-reflection* true)
 
 (defn format-sql
   "Return a nicely-formatted version of a `query` string with the current application db driver formatting."
   [sql]
-  (driver/prettify-native-form (mdb.connection/db-type) sql))
+  (driver/prettify-native-form (mdb/db-type) sql))
+
+(def ^:private NamespacedKeyword
+  [:and :keyword [:fn (comp seq namespace)]])
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply src/metabase/db/query.clj:46]: that's true but there are implications about whether it needs a nice error message. I'm going to punt on that and let another pass clean it up.

File: src/metabase/db.clj (hunk 2 above)
Comment (dpsutton) [reply src/metabase/db.clj]: fixed

File: src/metabase/db.clj (hunk 3 above)
Comment (dpsutton) [reply src/metabase/db.clj]: done! thanks

File: src/metabase/db.clj (hunk 4 above)
Comment (dpsutton) [reply src/metabase/db.clj]: fixed

File: src/metabase/db.clj (hunk 5 above)
Comment (dpsutton) [reply src/metabase/db.clj]: nice spot. thanks. fixed

File: enterprise/backend/src/metabase_enterprise/audit_app/pages/common.clj [hunk 8]
```diff
@@ -92,7 +91,7 @@
                   sql-jdbc.sync/db-default-timezone
                   :ttl/threshold (u/hours->ms 1))]
     (fn []
-      (timezone (mdb/db-type) {:datasource mdb.connection/*application-db*}))))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply enterprise/backend/src/metabase_enterprise/audit_app/pages/common.clj:95]: `*application-db*` is a `DataSource`

File: src/metabase/db/query.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/db/query.clj:46]: You can just use `clojure.core/qualified-keyword?` directly as a schema and it will do the right thing.

```clj
(malli.error/humanize (malli.core/explain qualified-keyword? :abc))
;; => ["should be a qualified keyword"]
```

If you write a `:fn` schema, you should include an :error/message` in the options, otherwise if that check fails it's going to give you a meaningless explanation:

```clj
(malli.error/humanize (malli.core/explain NamespacedKeyword :abc))
;; => ["unknown error"]

(def ^:private NamespacedKeyword2
  [:and :keyword [:fn {:error/message "should be a namespaced keyword"} (comp seq namespace)]])

(malli.error/humanize (malli.core/explain NamespacedKeyword2 :abc))
;; => ["should be a namespaced keyword"]
```
